// officeparser/dist/generators/BaseGenerator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseGenerator = void 0;
const configUtils_js_1 = require("../utils/configUtils.js");
const errorUtils_js_1 = require("../utils/errorUtils.js");
const styleMapper_js_1 = require("../utils/styleMapper.js");
/**
 * Base class for all document generators.
 * Provides common traversal logic and configuration handling.
 */
class BaseGenerator {
    destination;
    config;
    ast;
    messages = [];
    styleMapper;
    collectedNotes = [];
    constructor(destination, ast, config) {
        this.destination = destination;
        this.config = (0, configUtils_js_1.resolveGeneratorConfig)(destination, ast.config, config);
        this.ast = ast;
        this.styleMapper = new styleMapper_js_1.StyleMapper(this.config.styleMap, this.config.ignoreDefaultStyleMap);
    }
    /**
     * Retrieves the semantic mapping for a node, respecting the includeFormatting flag.
     * Per design requirements: Style mapping is bypassed if formatting is disabled.
     */
    getSemanticMapping(node) {
        if (this.config.includeFormatting === false) {
            return undefined;
        }
        return this.styleMapper.getMapping(node);
    }
    /**
     * Centralized logic for handling the onNode callback.
     * Evaluates the callback and returns a result that tells the generator how to proceed.
     *
     * @returns
     * - `string`: Use this as the node's output, skip default processing.
     * - `false`: Skip this node and its subtree.
     * - `void`: Proceed with default processing.
     */
    async handleOnNode(node) {
        const result = await this.config.onNode(node);
        if (result === false)
            return false;
        if (typeof result === 'string')
            return result;
    }
    /**
     * Recursively processes nodes and builds output.
     *
     * @param node - The current node being processed
     * @param processor - A function that takes a node and its children's output and returns the node's output string.
     * @returns The generated string for this node and its subtree.
     */
    async processNodeRecursive(node, processor) {
        const override = await this.handleOnNode(node);
        if (override === false)
            return '';
        if (typeof override === 'string')
            return override;
        let childrenOutput = '';
        if (node.children) {
            for (const child of node.children) {
                childrenOutput += await this.processNodeRecursive(child, processor);
            }
        }
        if (node.notes && node.notes.length > 0) {
            if (node.type !== 'slide') {
                this.collectedNotes.push(...node.notes);
            }
        }
        let result = await processor(node, childrenOutput);
        if (node.type === 'slide' && node.notes && node.notes.length > 0) {
            for (const note of node.notes) {
                result += await this.processNodeRecursive(note, processor);
            }
        }
        return result;
    }
    /**
     * Helper to generate a unique ID (slug) from text.
     */
    slugify(text) {
        return text
            .toLowerCase()
            .replace(/[^\w\s-]/g, '')
            .replace(/[\s_-]+/g, '-')
            .replace(/^-+|-+$/g, '');
    }
    /**
     * Recursively extracts plain text from a node and its children.
     */
    getNodeText(node) {
        if (node.text)
            return node.text;
        if (node.children) {
            return node.children.map(c => this.getNodeText(c)).join('');
        }
        return '';
    }
    /**
     * Reports a warning to the user and collects it for the final result.
     */
    warn(type, info, node) {
        const message = (0, errorUtils_js_1.getWarningMessage)(type, info);
        const issue = {
            type: 'warning',
            code: type,
            message,
            node,
            details: info
        };
        this.messages.push(issue);
        this.config.onWarning(issue);
    }
}
exports.BaseGenerator = BaseGenerator;
